sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryStatisticsPage.scala

```diff
@@ -199,49 +201,99 @@ private[ui] class StreamingQueryStatisticsPage(parent: StreamingQueryTab)
+          <td class={"aggregated-num-total-state-rows-histogram"}>{graphUIDataForNumberTotalRows.generateHistogramHtml(jsCollector)}</td>
         </tr>
-      // scalastyle:on
+          <tr>
+            <td style="vertical-align: middle;">
+              <div style="width: 160px;">
+                <div><strong>Aggregated Number Of Updated State Rows {SparkUIUtils.tooltip("Aggregated number of updated state rows.", "right")}</strong></div>
+              </div>
+            </td>
+            <td class={"aggregated-num-updated-state-rows-timeline"}>{graphUIDataForNumberUpdatedRows.generateTimelineHtml(jsCollector)}</td>
+            <td class={"aggregated-num-updated-state-rows-histogram"}>{graphUIDataForNumberUpdatedRows.generateHistogramHtml(jsCollector)}</td>
+          </tr>
+          <tr>
+            <td style="vertical-align: middle;">
+              <div style="width: 160px;">
+                <div><strong>Aggregated State Memory Used In Bytes {SparkUIUtils.tooltip("Aggregated state memory used in bytes.", "right")}</strong></div>
+              </div>
+            </td>
+            <td class={"aggregated-state-memory-used-bytes-timeline"}>{graphUIDataForMemoryUsedBytes.generateTimelineHtml(jsCollector)}</td>
+            <td class={"aggregated-state-memory-used-bytes-histogram"}>{graphUIDataForMemoryUsedBytes.generateHistogramHtml(jsCollector)}</td>
+          </tr>
+          <tr>
+            <td style="vertical-align: middle;">
+              <div style="width: 160px;">
+                <div><strong>Aggregated Number Of State Rows Dropped By Watermark {SparkUIUtils.tooltip("Aggregated number of state rows dropped by watermark.", "right")}</strong></div>
+              </div>
+            </td>
+            <td class={"aggregated-num-state-rows-dropped-by-watermark-timeline"}>{graphUIDataForNumRowsDroppedByWatermark.generateTimelineHtml(jsCollector)}</td>
+            <td class={"aggregated-num-state-rows-dropped-by-watermark-histogram"}>{graphUIDataForNumRowsDroppedByWatermark.generateHistogramHtml(jsCollector)}</td>
+          </tr>
+        // scalastyle:on
+
+      result ++= generateAggregatedCustomMetrics(query, minBatchTime, maxBatchTime, jsCollector)
+      result
     } else {
       new NodeBuffer()
     }
   }
 
+  def generateAggregatedCustomMetrics(
+     query: StreamingQueryUIData,
+     minBatchTime: Long,
+     maxBatchTime: Long,
+     jsCollector: JsCollector): NodeBuffer = {
+    val result: NodeBuffer = new NodeBuffer
+
+    // This is made sure on caller side but put it here to be defensive
+    require(query.lastProgress.stateOperators.nonEmpty)
+    val enabledCustomMetrics = parent.parent.conf.get(ENABLED_STREAMING_UI_CUSTOM_METRIC_LIST)
+    logDebug(s"Enabled custom metrics: $enabledCustomMetrics")
+    query.lastProgress.stateOperators.head.customMetrics.keySet().asScala
```

Comment:
Does this work if there are multiple state operators in the query plan and they have different custom metrics?

Reply:
This is working because all state operator must have the exact same number of custom metrics. The names are coming from head and the values from `query.recentProgress`.

Reply:
One additional thing, please see that the timeline and histogram aggregated...

Reply:
> This is working because all state operator must have the exact same number of custom metrics. 

@HeartSaVioR you added `customMetrics` to `StateOperatorProgress` in #21469 .
Did you intended that all the state operators in a query should have the same number / type of custom metrics?
In other words, each state operator should not have its own custom metrics right?

Reply:
@sarutak nice catch, that's definitely not tested and the code will blow up.
@HeartSaVioR If operator `O1` has custom metric `C1` and operator `O2` has custom metric `C2` then the currently implemented aggregation is not fine.

Reply:
Solved all other comments but leaving this open because this can be a major issue w/ big influence on the PR.
Let's wait on the answer and adjust the code/tests based on that (as said this case not tested because of maybe bad assumption).

Reply:
The custom metrics is bound to the `state store provider` instead of `operator`.

Given the provider instances are created per state store id, it's not technically impossible for provider to provide different custom metrics, but at least that's not a purpose and you'll wonder how to leverage the flexibility. These metrics are for state itself, not related to the operator.

Reply:
`StateStoreWriter#getProgress` is not marked as `final` so I wondered it's intended to have sub classes which override `getProgress` where custom metrics can be created for a operator which is implemented as a sub class of `StateStoreWriter`.
But now I understand that custom metrics are not for operators.

Reply:
OK you found a good point I was missing.

https://github.com/apache/spark/blob/5af5aa146ecbff38b809127b5eb9805441627ed2/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala#L127-L137

Here the list of custom metrics are being picked up, and it is populated from dummy StateStoreProvider instance. That said, Spark "expects" the same value of `supportedCustomMetrics` across StateStoreProvider instances created from same provider class.

Reply:
Thank you guys pinpointing this + giving explanation related the main intention of this feature.
As I understand the mentioned use-case is physically possible but that's not the main intention.
What is the final call what satisfies everybody? Should we make `StateStoreWriter#getProgress` as `final` to enforce this or we're good as-is?

Reply:
How about leaving it as is and focusing the UI for now?

Reply:
I agree w/ this! Thanks @sarutak